adapters/bidmachine/bidmachine.go

```diff
@@ -0,0 +1,223 @@
+
+	request.Imp = impressions
+
+	if len(result) == 0 {
+		return nil, errs
+	}
+	return result, errs
+}
+
+func hasRewardedBattr(attr []openrtb.CreativeAttribute) bool {
+	for i := 0; i < len(attr); i++ {
+		if attr[i] == openrtb.CreativeAttribute(16) {
+			return true
+		}
+	}
+	return false
+}
+
+func (a *BidmachineAdapter) MakeBids(request *openrtb.BidRequest, _ *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	var errs []error
+
+	switch responseData.StatusCode {
+	case http.StatusNoContent:
+		fallthrough
+	case http.StatusServiceUnavailable:
+		return nil, nil
+
+	case http.StatusBadRequest:
+		fallthrough
+	case http.StatusUnauthorized:
+		fallthrough
+	case http.StatusForbidden:
+		return nil, []error{&errortypes.BadInput{
+			Message: "unexpected status code: " + strconv.Itoa(responseData.StatusCode) + " " + string(responseData.Body),
+		}}
+
+	case http.StatusOK:
+		break
+
+	default:
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: "unexpected status code: " + strconv.Itoa(responseData.StatusCode) + " " + string(responseData.Body),
+		}}
+	}
+
+	var bidResponse openrtb.BidResponse
+	err := json.Unmarshal(responseData.Body, &bidResponse)
+	if err != nil {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: err.Error(),
+		}}
+	}
+
+	response := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+
+	for _, seatBid := range bidResponse.SeatBid {
+		for _, bid := range seatBid.Bid {
+			thisBid := bid
+			bidType := GetMediaTypeForImp(bid.ImpID, request.Imp)
+			if bidType == UndefinedMediaType {
+				errs = append(errs, &errortypes.BadServerResponse{
+					Message: "ignoring bid id=" + bid.ID + ", request doesn't contain any valid impression with id=" + bid.ImpID,
+				})
+				continue
+			}
+			response.Bids = append(response.Bids, &adapters.TypedBid{
+				Bid:     &thisBid,
+				BidType: bidType,
+			})
+		}
+	}
+
+	return response, errs
+}
+
+// Builder builds a new instance of the Bidmachine adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	template, err := template.New("endpointTemplate").Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
+	}
+
+	bidder := &BidmachineAdapter{
+		endpoint: *template,
+	}
+
+	return bidder, nil
+}
+
+const UndefinedMediaType = openrtb_ext.BidType("")
+
+func (a *BidmachineAdapter) buildEndpointURL(params openrtb_ext.ExtImpBidmachine) (string, error) {
+	endpointParams := macros.EndpointTemplateParams{Host: params.Host}
+	return macros.ResolveMacros(a.endpoint, endpointParams)
+}
+
+func GetMediaTypeForImp(impID string, imps []openrtb.Imp) openrtb_ext.BidType {
+	var bidType openrtb_ext.BidType = UndefinedMediaType
+	for _, impression := range imps {
+		if impression.ID != impID {
+			continue
+		}
+		switch {
+		case impression.Banner != nil:
```

Comment:
This is an anti-pattern. Does your server respond with the type of the creative via ext fields?